### PR TITLE
Hotfix double destroy

### PR DESCRIPTION
**Pull request by jkeon**
If a `GameObject` is destroyed via GameObject.Destroy, it will then trigger the Dispose which will then trigger the Destroy again. 

This was discovered by having an `AbstractAnvilMonoBehaviour` executing in the Editor via the `ExecuteAlways` attribute and seeing how Unity would call Destroy on it resulting in the double destroy. 

### What issues does this resolve?
- None

### What PRs does this depend on?
 - None

### Does this introduce a breaking change?
 - [ ] Yes 
 - [x] No

### CODE REVIEW

File: Scripts/Runtime/Core/AbstractAnvilMonoBehaviour.cs [hunk 1]
```diff
@@ -58,17 +60,39 @@ public void Dispose()
...
             IsDisposing = true;
             DisposeSelf();
+
+            if (!m_IsDestroying)
+            {
+#if !UNITY_EDITOR
             Destroy(gameObject);
+#else
+                //If we're running in the Editor via the ExecuteAlways attribute, we're not allowed to call Destroy
+                if (Application.isEditor
```
[the remark this answers is not in the document]
Comment (jkeon) [reply Scripts/Runtime/Core/AbstractAnvilMonoBehaviour.cs]: Good point, removed.

File: Scripts/Runtime/Core/AbstractAnvilMonoBehaviour.cs [hunk 2]
```diff
@@ -58,17 +60,39 @@ public void Dispose()
...
-            if (IsDisposing || IsDisposed)
+            if (IsDisposing
+             || IsDisposed
+             || m_IsDestroying)
             {
                 return;
             }
+            m_IsDestroying = true;
 
             Dispose();
```
[the remark this answers is not in the document]
Comment (jkeon) [reply Scripts/Runtime/Core/AbstractAnvilMonoBehaviour.cs]: Invalid because your overall suggestion was a better idea.

**Comment by jkeon** [comment]
> I've left a couple comments on the current implementation but I'm wondering if we'd be better served by changing up the pattern a bit to avoid needing the extra field.
> 
> If we change the signature of Dispose to Dispose(bool skipDestroy = false) then we could bypass the destroy call when coming from the OnDestroy() stack without the need to carry a field on the type.

Yeah this was a better idea. Changed to reflect that. We still have the original signature so we match the `IDisposable` interface but then we just pipe to the param version.

**Review by mbaker3 — APPROVED** [review]

File: Scripts/Runtime/Core/AbstractAnvilMonoBehaviour.cs [hunk 3]
```diff
@@ -50,6 +50,11 @@ protected virtual void Awake()
         /// Will Destroy the attached <see cref="GameObject"/> if it has not been destroyed yet.
         /// </summary>
         public void Dispose()
+        {
+            Dispose(true);
+        }
+
+        private void Dispose(bool shouldDestroyGameObject)
```
Comment (mbaker3) [inline Scripts/Runtime/Core/AbstractAnvilMonoBehaviour.cs:57]: Might make sense to make this public some day if you only want to dispose the specific behaviour but we should wait until the need actually comes up.

We'd need a `RemoveComponent` path anyway.

Comment (jkeon) [reply Scripts/Runtime/Core/AbstractAnvilMonoBehaviour.cs:57]: Yeah that's fair. Interesting that in all these years that's never been a pattern we've done.

Comment (mbaker3) [reply Scripts/Runtime/Core/AbstractAnvilMonoBehaviour.cs:57]: Yea, the whole dispose pattern is a bit of an opinionated choice.

On one hand you could say that we should have `Dispose` at all and developers should just use the engine commands `Object.Destroy` and `MonoBehaviour.RemoveComponent`.

On the other hand, it's nice to implement the `IDisposable` pattern and, personally, I prefer to call `Dispose` vs using Unity's various ways of destroying objects.

Comment (jkeon) [reply Scripts/Runtime/Core/AbstractAnvilMonoBehaviour.cs:57]: Agreed!

**Pull request merged**